subprojects/core/src/main/java/org/gradle/execution/plan/FinalizerGroup.java

```diff
@@ -260,21 +265,26 @@ private MemberSuccessors getNodesThatBlock(Node node) {
         return successors.getNodesThatBlock(node);
     }
 
-    private boolean dependsOn(Node fromNode, Node toNode) {
-        Set<Node> seen = new HashSet<>();
-        List<Node> queue = new ArrayList<>();
-        Iterables.addAll(queue, fromNode.getHardSuccessors());
-        while (!queue.isEmpty()) {
-            Node node = queue.remove(0);
-            if (node == toNode) {
-                return true;
-            }
-            if (!seen.add(node)) {
-                continue;
+    private Set<Node> getDependenciesThatAreMembers(Set<Node> blockedFinalizedMembers) {
+        Set<Node> dependenciesThatAreMembers = new HashSet<>(members.size());
+        Set<Node> seen = new HashSet<>(1024);
+
+        for (Node fromNode : blockedFinalizedMembers) {
+            ArrayDeque<Node> queue = new ArrayDeque<>(1024);
```

Comment:
🚀 Would it make sense to allocate the queue only once outside the loop?

Reply:
Yep, that makes sense.  Done.